beacon-chain/casper/sharding.go

```diff
@@ -7,51 +7,67 @@ import (
-		committeesPerSlot = len(indices)/params.CycleLength/(params.MinCommiteeSize*2) + 1
-		slotsPerCommittee = 1
-	} else {
-		committeesPerSlot = 1
-		slotsPerCommittee = 1
-		for len(indices)*slotsPerCommittee < params.MinCommiteeSize && slotsPerCommittee < params.CycleLength {
-			slotsPerCommittee *= 2
-		}
-	}
 
 	// split the shuffled list for heights.
-	shuffledList, err := utils.ShuffleIndices(seed, indices)
+	shuffledValidators, err := utils.ShuffleIndices(seed, indices)
 	if err != nil {
 		return nil, err
 	}
 
-	heightList := utils.SplitIndices(shuffledList, params.CycleLength)
+	return splitBySlotShard(shuffledValidators, crosslinkStartShard), nil
+}
+
+// splitBySlotShard splits the validator list into evenly sized committees and assigns each
+// committee to a slot and a shard. If the validator set is large, multiple committees are assigned
+// to a single slot and shard. If the validator set is small, a single committee is assigned to a shard
+// across multiple slots. See getCommitteeParams for more details.
+func splitBySlotShard(shuffledValidators []uint32, crosslinkStartShard uint64) []*pb.ShardAndCommitteeArray {
+	committeesPerSlot, slotsPerCommittee := getCommitteeParams(len(shuffledValidators))
+
+	committeBySlotAndShard := []*pb.ShardAndCommitteeArray{}
 
 	// split the shuffled height list for shards
-	for i, subList := range heightList {
-		shardList := utils.SplitIndices(subList, params.MinCommiteeSize)
-		for _, shardIndex := range shardList {
-			shardID := int(crosslinkStartShard) + i*committeesPerSlot/slotsPerCommittee
-			committees = append(committees, &BeaconCommittee{
-				ShardID:   shardID,
-				Committee: shardIndex,
+	validatorsBySlot := utils.SplitIndices(shuffledValidators, params.CycleLength)
+	for i, validatorsForSlot := range validatorsBySlot {
+		shardCommittees := []*pb.ShardAndCommittee{}
+		validatorsByShard := utils.SplitIndices(validatorsForSlot, committeesPerSlot)
+		shardStart := int(crosslinkStartShard) + i*committeesPerSlot/slotsPerCommittee
+		for j, validatorsForShard := range validatorsByShard {
+			shardID := (shardStart + j) % params.ShardCount
+			shardCommittees = append(shardCommittees, &pb.ShardAndCommittee{
+				ShardId:   uint64(shardID),
+				Committee: validatorsForShard,
 			})
 		}
+
+		committeBySlotAndShard = append(committeBySlotAndShard, &pb.ShardAndCommitteeArray{
+			ArrayShardAndCommittee: shardCommittees,
+		})
+	}
+
+	return committeBySlotAndShard
+}
+
+// getCommitteeParams calculates the parameters for ShuffleValidatorsToCommittees.
+// If numActiveValidators > CycleLength * MinCommitteeSize, multiple committees are selected
+// to attest the same shard in a single slot.
+// If numActiveValidators < CycleLength * MinCommitteeSize, committees span across multiple slots
+// to attest the same shard.
+func getCommitteeParams(numValidators int) (committeesPerSlot, slotsPerCommittee int) {
+	if numValidators >= params.CycleLength*params.MinCommiteeSize {
```

Comment:
return the committeesperslot and slotspercommittee here instead of going into an else branch for greater readability

Reply:
good point :+1: